File: Sources/Server/Sources/Game/Entity/Dynamics/Mob/MobSpecialMovementSinusodialWave.ts
const TAU = Math.PI * 2;

function generateSinWaveTable(sampleRate: number) {
    const samples = new Array(sampleRate);

    for (let i = 0; i < sampleRate; i++) {
        const y = Math.sin((i / sampleRate) * TAU);

        samples[i] = y;
    }

    return samples;
}

export default class SinusodialWave {
    private static table: Array<number>;
    static {
        this.table = generateSinWaveTable(100);
    }

    public static at(t: number): number {
        return this.table[t % this.table.length];
    }
}

File: Sources/goserver/internal/wave/mob_special_movement_sinusodial_wave.go
package wave

import "math"

const TAU = math.Pi * 2

func generateSinWaveTable(sampleRate int) []float64 {
    samples := make([]float64, sampleRate)
    
    for i := range sampleRate {
        y := math.Sin(float64(i) / float64(sampleRate) * TAU)
        samples[i] = y
    }
    
    return samples
}

type SinusoidalWave struct {
    table []float64
}

func NewSinusoidalWave(sampleRate int) *SinusoidalWave {
    return &SinusoidalWave{
        table: generateSinWaveTable(sampleRate),
    }
}

func (w *SinusoidalWave) At(t int) float64 {
    index := t % len(w.table)
    return w.table[index]
}

var SinusodialWave = NewSinusoidalWave(200)